site/content/docs/user/exposing-applications.md

```diff
@@ -0,0 +1,261 @@
+kind-worker          Ready    <none>   5m58s   v1.14.1   172.17.0.2    <none>        Ubuntu Disco Dingo (development branch)   4.9.125-linuxkit   containerd://1.2.6-0ubuntu1
+kind-worker2         Ready    <none>   5m59s   v1.14.1   172.17.0.3    <none>        Ubuntu Disco Dingo (development branch)   4.9.125-linuxkit   containerd://1.2.6-0ubuntu1
+```
+
+### Kubectl port-forward
+
+This is the most simple one and well documented
+
+https://kubernetes.io/docs/tasks/access-application-cluster/port-forward-access-application-cluster/
+
+### Port forwarding with socat
+
+Docker allows to port map containers ports to our localhost, the problem is that this port mapping can be configured only when we create the container. If we want to map new port we have to recreate the container.
+
+We can leverage this functionality to create one container that allows to forward ports, this container will forward the new ports mapped to the corresponding ports on the cluster.
+
+This process is well documented on different places, ie.
+
+https://sosedoff.com/2018/04/25/expose-docker-ports.html
+
+https://andrewwilkinson.wordpress.com/2017/11/22/exposing-docker-ports-after-the-fact/
+
+In our example we can connect to the exposed nodePort in one of the cluster nodes:
+
+```sh
+docker run --rm -d -p80:8080 --name=proxy alpine/socat TCP4-LISTEN:8080,fork,reuseaddr TCP4:172.17.0.4:30039
+```
+
+and access the service in our host:
+
+```sh
+$ curl localhost
+Hello Kubernetes!
+```
+
+
+### Socks Proxy
+
+> A SOCKS server is a general purpose proxy server that establishes a TCP connection to another server on behalf of a client, then routes all the traffic back and forth between the client and the server. It works for any kind of network protocol on any port
+
+Instead of having to recreate the port mapping constantly we can create a socks proxy inside the docker network and use it to connect to the exposed applications.
+
+SOCKS proxy can be used with any web browser, just have to configure your application to use it and use your service IP and Port target directly.
+
+Create a container with a SOCKS proxy and forward the port
+
+```sh
+docker run --rm -d --cap-add=NET_ADMIN --name socks5 -p 1080:1080 aojea/socks5
+```
+
+Accessing the NodePort service, for that we need only the IP of one of the cluster nodes:
+
+```sh
+curl --socks5 localhost:1080 172.17.0.4:30039
+Hello Kubernetes!
+```
+
+We can reach the service IP too, for that we have to add the routes to the service Subnet through one of the cluster nodes. By default `kubeadm` uses "10.96.0.0/12" but it can changed.
+
+```sh
+docker exec -it socks5 sh
+# add the ip route through the kind-control-plane node
+ip route add 10.96.0.0/12 via 172.17.0.4
+```  
+
+And now we can access the application using the the Service IP and Service port:
+
+```sh
+curl --socks5 localhost:1080  10.96.59.230:8080
+Hello Kubernetes!
+```
+
+You can use other browser like Google Chrome or Mozilla Firefox too.
```

Comment:
maybe state that the browser would need to be configured with socks proxy, and maybe links to a doc for each browser mentioned that explains the details.